sway-lib-std/src/alloc.sw

```diff
@@ -0,0 +1,45 @@
+//! Library for allocating memory
+//! Inspired from: https://doc.rust-lang.org/std/alloc/index.html
+library alloc;
+
+use ::intrinsics::*;
+
+/// Allocates zeroed memory on the heap
+/// 
+/// In FuelVM, the heap begins at `VM_MAX_RAM - 1` and grows downward.
+/// Heap pointer `$hp` will always point to unallocated space.
+/// 
+/// Initially the heap will look like this:
+/// ... 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 |
+///                                               $hp^  ^VM_MAX_RAM
+/// 
+/// After allocating with `let ptr = alloc(8)`:
+/// ... 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 |
+///                       $hp^  ^ptr                    ^VM_MAX_RAM
+/// 
+/// After writing with `sw(ptr, u64::max())`:
+/// ... 00 00 00 00 00 00 00 00 FF FF FF FF FF FF FF FF |
+///                       $hp^  ^ptr                    ^VM_MAX_RAM
+///
+/// See: https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/main.md#vm-initialization
+/// See: https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/opcodes.md#aloc-allocate-memory
```

Comment:
OK, I believe the initialisation of the VM is broken, and `$hp` should be initialised to `VM_MAX_RAM`.  Putting in hacks to account for the neglected last byte isn't the role of the stdlib.

Especially since if another part of the library used `ALOC` and didn't do the little `$hp + 1` trick then there'd be conflict and potential corruption.